File: e2e/testdata/sdk_driver.py
"""e2e driver for the Python SDK (run as a real subprocess by e2e/sdk_test.go).

usage: sdk_driver.py SPOOL CHAIN GOAL START COUNT FLUSH_TIMEOUT [lose-first-ack]
Records COUNT records (payload {"n": START+i}) and flushes; prints "pending=<n>".
With lose-first-ack the first successful POST's response is dropped (as if the connection died
after ledgerd committed), so the SDK re-sends it and only the idempotency key prevents a duplicate.
"""
import sys

from ledger_sdk import Ledger

spool, chain, goal, start, count, flush_timeout = sys.argv[1:7]
lose = len(sys.argv) > 7 and sys.argv[7] == "lose-first-ack"
led = Ledger(chain, spool_path=spool, flush_on_exit=False, base_retry_delay=0.05, max_retry_delay=0.5, timeout=2.0)
if lose:
    real = led._transport
    state = {"lost": False}

    def lossy(path, body):
        res = real(path, body)
        if not state["lost"]:
            state["lost"] = True
            raise ConnectionResetError("e2e: response lost after commit")
        return res

    led._transport = lossy
actors = [{"kind": "human", "id": "alice"}, {"kind": "agent", "id": "py-agent", "model": "m", "model_version": "1"}]
for i in range(int(start), int(start) + int(count)):
    led.record("ledger.memory.written", {"key": "n", "n": i, "sdk": "python"}, actors, goal_id=goal)
print("queued", flush=True)
led.flush(timeout=float(flush_timeout))
led.close()
print("pending=%d" % led.pending(), flush=True)

File: e2e/testdata/sdk_driver.mjs
// e2e driver for the TypeScript SDK (see sdk_driver.py for the protocol).
// usage: node sdk_driver.mjs SDK_DIST SPOOL CHAIN GOAL START COUNT FLUSH_TIMEOUT_S [lose-first-ack]
const [dist, spool, chain, goal, start, count, flushTimeout, mode] = process.argv.slice(2);
const { Ledger } = await import(dist + "/index.js");
let lost = false;
const lossy = async (url, init) => {
  const res = await fetch(url, init);
  if (mode === "lose-first-ack" && res.ok && !lost) {
    lost = true;
    throw new TypeError("e2e: response lost after commit");
  }
  return res;
};
const led = new Ledger({
  chain, spoolPath: spool, flushOnExit: false, baseRetryDelayMs: 50, maxRetryDelayMs: 500, timeoutMs: 2000, fetch: lossy,
});
const actors = [{ kind: "human", id: "alice" }, { kind: "agent", id: "ts-agent", model: "m", model_version: "1" }];
for (let i = Number(start); i < Number(start) + Number(count); i++) {
  led.record("ledger.memory.written", { key: "n", n: i, sdk: "ts" }, actors, { goalId: goal });
}
console.log("queued");
await led.flush(Number(flushTimeout) * 1000);
led.close();
console.log(`pending=${led.pending}`);
process.exit(0);
